python-checks/src/main/java/org/sonar/python/checks/HardCodedCredentialsCheck.java

```diff
@@ -0,0 +1,292 @@
+      ctx.addIssue(stringLiteral, MESSAGE);
+    }
+  }
+
+  private static boolean isURLWithCredentials(StringLiteral stringLiteral) {
+    try {
+      URL url = new URL(stringLiteral.trimmedQuotesValue());
+      String userInfo = url.getUserInfo();
+      if (userInfo != null && userInfo.matches("\\S+:\\S+")) {
+        return true;
+      }
+    } catch (MalformedURLException e) {
+      return false;
+    }
+    return false;
+  }
+
+  private void handleBinaryExpression(BinaryExpression binaryExpression, SubscriptionContext ctx) {
+    boolean shouldReport = false;
+    if (binaryExpression.leftOperand() instanceof HasSymbol && binaryExpression.rightOperand().is(Tree.Kind.STRING_LITERAL)) {
+      shouldReport = isSymbolCredential(((HasSymbol) binaryExpression.leftOperand()).symbol());
+    }
+    if (binaryExpression.rightOperand() instanceof HasSymbol && binaryExpression.leftOperand().is(Tree.Kind.STRING_LITERAL)) {
+      shouldReport = isSymbolCredential(((HasSymbol) binaryExpression.rightOperand()).symbol());
+    }
+    if (shouldReport) {
+      ctx.addIssue(binaryExpression, MESSAGE);
+    }
+  }
+
+  private void handleAssignmentStatement(AssignmentStatement assignmentStatement, SubscriptionContext ctx) {
+    ExpressionList lhs = assignmentStatement.lhsExpressions().get(0);
+    Expression expression = lhs.expressions().get(0);
+
+    if (expression instanceof HasSymbol) {
+      Symbol symbol = ((HasSymbol) expression).symbol();
+      if (isSymbolCredential(symbol)) {
+        checkAssignedValue(assignmentStatement, ctx);
+      }
+    }
+
+    if (expression.is(Kind.SUBSCRIPTION)) {
+      SubscriptionExpression subscriptionExpression = (SubscriptionExpression) expression;
+      if (subscriptionExpression.subscripts().expressions().stream().anyMatch(e -> e.is(Kind.STRING_LITERAL) &&
+        isCredential(((StringLiteral) e).trimmedQuotesValue(), variablePatterns()))) {
+        checkAssignedValue(assignmentStatement, ctx);
+      }
+    }
+  }
+
+  private void checkAssignedValue(AssignmentStatement assignmentStatement, SubscriptionContext ctx) {
+    if (assignmentStatement.assignedValue().is(Kind.STRING_LITERAL)) {
+      StringLiteral literal = (StringLiteral) assignmentStatement.assignedValue();
+      if (!isEmptyStringLiteral(literal) && !isCredential(literal.trimmedQuotesValue(), variablePatterns())) {
+        ctx.addIssue(assignmentStatement, MESSAGE);
+      }
+    }
+  }
+
+  private boolean isSymbolCredential(@CheckForNull Symbol symbol) {
+    return symbol != null && isCredential(symbol.name(), variablePatterns());
+  }
+
+  private static boolean isNonEmptyStringLiteral(Tree tree) {
+    return tree.is(Kind.STRING_LITERAL) && !isEmptyStringLiteral((StringLiteral) tree);
+  }
+
+  private static boolean isEmptyStringLiteral(StringLiteral stringLiteral) {
+    return stringLiteral.trimmedQuotesValue().isEmpty();
+  }
+
+  private static boolean isCredential(String target, Stream<Pattern> patterns) {
```

Comment:
I think I'd refactor `isSymbolCredential` and `isCredential` in a single method:
```java
private static boolean isCredential(Tree tree, Stream<Pattern> patterns) { 
  if (tree instanceof HasSymbol) { 
     // ... 
  } else if (tree.is(Kind.STRING_LITERAL)) {
    // ... 
  }
} 
```
However feel free to discard this comment, it's just a personal preference. :)